internal/dialer/dialer.go

```diff
@@ -0,0 +1,256 @@
+	ctx context.Context, network, address string) (net.Conn, error) {
+	ev := Events{Network: network, RemoteAddress: address, StartTime: time.Now()}
+	conn, err := d.Dialer.DialContext(ctx, network, address)
+	ev.EndTime = time.Now()
+	ev.Error = err
+	if conn != nil {
+		ev.LocalAddress = conn.LocalAddr().String()
+	}
+	d.mu.Lock()
+	d.events = append(d.events, ev)
+	d.mu.Unlock()
+	return conn, err
+}
+
+// Events contains events generated when dialing
+type Events struct {
+	RemoteAddress string
+	EndTime       time.Time
+	Error         error
+	LocalAddress  string
+	Network       string
+	StartTime     time.Time
+}
+
+// Service is the dialing service. This is useful to create a flexible HTTP
+// transport, for which the dialer is undertermined. Just pass an instance of
+// this struct to httptransport.NewBase(). Every dial request will block and
+// possibly timeout, until a single goroutine takes ownership of this instance
+// of Service by running service.Loop. Because service.Loop takes in input
+// a dialer, it means you can customize dial depending on circumstances.
+type Service struct {
+	ch   chan serviceRequest
+	sync chan interface{}
+}
+
+// NewService creates a new dialing service.
+func NewService() Service {
+	return Service{ch: make(chan serviceRequest), sync: make(chan interface{}, 1)}
+}
+
+type serviceRequest struct {
+	address string
+	conn    chan net.Conn
+	err     chan error
+	network string
+}
+
+// DialContext implements Dialer.DialContext.
+func (d Service) DialContext(
+	ctx context.Context, network, address string) (net.Conn, error) {
+	req := serviceRequest{
+		address: address,
+		conn:    make(chan net.Conn),
+		err:     make(chan error),
+		network: network,
+	}
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case d.ch <- req:
+	}
+	select {
+	case conn := <-req.conn:
+		return conn, nil
+	case err := <-req.err:
+		return nil, err
+	case <-ctx.Done():
+		return nil, ctx.Err() // close managed by s.Loop
+	}
+}
+
+// ErrBusy means that Service.Loop is already running
+var ErrBusy = errors.New("service loop is busy")
+
+// Loop runs the dialer.Service loop until ctx is done. If another instance of
+// the loop is running, this function returns ErrBusy. This guarantees that two
+// different subsystems cannot independently service dial requests. Instead a
+// single subsystem at a time will be the loop owner.
+func (d Service) Loop(ctx context.Context, dialer Dialer) error {
+	select {
+	case d.sync <- true:
+	default:
+		return ErrBusy
+	}
+	defer func() { <-d.sync }()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil // okay to starve requests because they have ctx
+		case req := <-d.ch:
+			conn, err := dialer.DialContext(ctx, req.network, req.address)
+			if err != nil {
+				select {
+				case req.err <- err:
+				default:
+				}
+			} else {
+				select {
+				case req.conn <- conn:
+				default:
+					conn.Close() // cannot sent? we cleanup
+				}
+			}
+		}
+	}
+}
```

Comment:
Alternative, perhaps better implementation:

```Go
func (d Service) Start(ctx context.Context, dialer Dialer) error {
	select {
	case d.sync <- true:
	default:
		return ErrBusy
	}
	go func() {
		defer func() { <-d.sync }()
		for {
			select {
			case <-ctx.Done():
				return nil // okay to starve requests because they have ctx
			case req := <-d.ch:
				conn, err := dialer.DialContext(ctx, req.network, req.address)
				if err != nil {
					select {
					case req.err <- err:
					default:
					}
				} else {
					select {
					case req.conn <- conn:
					default:
						conn.Close() // cannot sent? we cleanup
					}
				}
			}
		}
	}()
	return nil
}
```

Reply:
So:

```Go
func (d Service) Start(ctx context.Context, dialer Dialer) error {
	select {
	case d.sync <- true:
	default:
		return ErrBusy
	}
	go func() {
		defer func() { <-d.sync }()
		d.loop(ctx, dialer)
	}()
	return nil
}

func (d Service) loop(ctx context.Context, dialer Dialer) {
	var wg sync.WaitGroup
	for {
		select {
		case <-ctx.Done():
			wg.Wait()
			return nil // okay to starve requests because they have ctx
		case req := <-d.ch:
			wg.Add(1)
			go func() {
				defer wg.Done()
				d.dial(ctx, dialer, req, &wg)
			}()
		}
	}
}

func (d Service) dial(
	ctx context.Context, dialer Dialer, req serviceRequest, wg *sync.WaitGroup) {
	conn, err := dialer.DialContext(ctx, req.network, req.address)
	if err != nil {
		select {
		case req.err <- err:
		default:
		}
	} else {
		select {
		case req.conn <- conn:
		default:
			conn.Close() // cannot sent? we cleanup
		}
	}
}
```